feat(postgres): Add PostgresCDCEventTrigger with AssetWatcher integration

jose-lehmkuhl opened the pull request:
Description
- Adds PostgresCDCEventTrigger to Postgres provider.
- Supports polling-based CDC using timestamp or version column.
- Integrates with AssetWatcher for triggering Airflow assets.
- Includes unit tests and example DAG.
- Adds state_key parameter for configurable state persistence.
- Updates provider.yaml and generated provider info.

All pre-commit checks and tests passed successfully.

Closes: none
Related: none

---
Notes
This feature enables simple polling-based CDC for PostgreSQL within Airflow, complementing the Asset-based workflow model introduced in Airflow 3.x.

This feature requires Airflow 3.0+ as it depends on `AssetWatcher` and `BaseEventTrigger`.

jason810496 commented on providers/postgres/src/airflow/providers/postgres/triggers/postgres_cdc.py:
```diff
@@ -0,0 +1,125 @@
+                    user=conn.login,
+                    password=conn.password,
+                ) as pg_conn:
+                    with pg_conn.cursor() as cursor:
```
How about using existed `PostgresHook` to establish connection ?

jose-lehmkuhl replied:
I've updated the code to use PostgresHook.get_conn() instead of psycopg2. Makes it cleaner and more consistent with Airflow patterns. Thanks!

jason810496 commented on providers/postgres/src/airflow/providers/postgres/triggers/postgres_cdc.py:
```diff
@@ -0,0 +1,125 @@
+            },
+        )
+
+    async def run(self) -> AsyncGenerator[TriggerEvent, None]:
```
```suggestion
    async def run(self) -> AsyncGenerator[TriggerEvent]:
```

jason810496 commented on providers/postgres/src/airflow/providers/postgres/triggers/postgres_cdc.py:
```diff
@@ -0,0 +1,125 @@
+                        result = cursor.fetchone()
+                        max_value: datetime.datetime | None = result[0]
+
+                        if max_value is not None:
```
I think we could flatten the `max_value` criteria with early return:
```python
if not max_value:
    self.log.info("No data found in %s for column %s.", self.table, self.cdc_column)
    return

# ... rest of logic
```

jose-lehmkuhl replied:
Done! I've refactored the condition with an early return to simplify the flow. Thanks for the suggestion!

jason810496 commented on providers/postgres/tests/system/postgres/example_postgres_cdc_asset_watcher.py:
```diff
@@ -0,0 +1,97 @@
+            host=conn.host, port=conn.port, dbname=conn.schema, user=conn.login, password=conn.password
+        ) as pg_conn:
+            with pg_conn.cursor() as cursor:
+                last_value = Variable.get(STATE_KEY, default_var=None)
```
How about having a `state_key` readonly property in `PostgresCDCEventTrigger`, so that user can get last value with:
```python
last_value = Variable.get(cdc_trigger.state_key, default_var=None)
```

jose-lehmkuhl replied:
I've added a @property for state_key so users can access it directly. Thanks for the suggestion!

vincbeck commented on providers/postgres/src/airflow/providers/postgres/triggers/postgres_cdc.py:
```diff
@@ -0,0 +1,125 @@
+
+                            if max_value > last_dt:
+                                self.log.info("New change detected: %s (iso: %s)", max_value, max_iso)
+                                yield TriggerEvent(
```
Interesting usage of Airflow variable to save the last read. However, you never save the new value in variables,  from the example below, you expect users to do it? I think you should do it here, right before sending the event.

jason810496 replied:
> Interesting usage of Airflow variable to save the last read. 

+1, I have never thought this scenario before

> I think you should do it here, right before sending the event.

Also agree with this

jose-lehmkuhl replied:
Correct me if I'm wrong, but from my understanding and testing, triggers in Airflow (running inside the triggerer process) don't have permission or capability to perform Variable.set() operations, as they primarily rely on the airflow.sdk client, which doesn't expose such functionality within the triggerer context.

I initially considered saving the new CDC value directly inside the trigger, but it seems that only Variable.get() is accessible in this context, and Variable.set() is either unsupported or leads to runtime issues.

I'm totally open to suggestions if there's a recommended approach for persisting state from within triggers, or if there's some mechanism to enable this safely. Thanks for the great feedback!

vincbeck replied:
Maybe @ashb (or others) can help us here

jroachgolf84 replied:
This is a conversation that @dstandish and I have also been having. It's something that I (we) believe will pop up quite a bit with Asset watchers...

jroachgolf84 replied:
I think this is a quick-and-dirty to get started, but I'm worried there might be some edge-cases here/confusion for users. Like mentioned above, what if the DAG author forgets to set the Variable? Or what if another developer deletes the Variable because they're not sure where it came from? What if there are multiple AssetWatchers for a single table? It's a tough question to answer...

jose-lehmkuhl replied:
Thanks for raising these important points!

I completely agree that relying on the DAG author to manage state persistence introduces risks of inconsistency, especially in scenarios with multiple AssetWatchers or when variables are deleted or forgotten.

In the current implementation, the state_key is explicitly configurable when initializing PostgresCDCEventTrigger. If multiple AssetWatchers are configured for the same table but with different state_key values, they will maintain independent state, avoiding direct conflicts. Of course, if users accidentally reuse the same state_key for multiple watchers, it could lead to unexpected behaviors.

Given the current design and behavior of triggers regarding state persistence, this approach seemed like a simple and native way to address the problem without adding complexity or requiring external solutions.

Perhaps there's room for a more integrated or standardized mechanism within Airflow's Asset/Trigger ecosystem to manage this kind of ephemeral or durable state safely. I'm happy to collaborate further on possible improvements in this area!

vincbeck replied:
Agree on that, I also think we can do better and implement a mechanism to share a state between triggerers. This would simplify this kind of use case.

jason810496 reviewed (commented):
Thanks for the PR! looks good overall.
Nice to see new Postgres provider integration.

eladkal commented on providers/postgres/docs/changelog.rst:
Please undo changes to this file.

jose-lehmkuhl replied:
Done! I've reverted the changes to changelog.rst. Thanks for pointing that out!

jose-lehmkuhl replied:
I've also added a note in the class docstring warning users about potential high event frequency and suggesting adjusting polling_interval accordingly. Thanks for raising this!

eladkal reviewed (commented):
CDC events can (and probably will) be very frequent. It can easily get to millions of events for high load databases. Should we place some kind of warning about resonable load?

vincbeck reviewed (commented):
Overall looks good. I still think this is more a temporary solution than a production-grade solution. Relying on variables and the fact that the user will update this variable to the latest timestamp will not work on the long run. We should definitely invest on implementing a robust solution to share a state among triggers. This is actually a super interesting topic, if anybody has some bandwith :) If someone is interested, I'll be happy to help.

Action item for this PR: can you please create docs? Explaining the trigger, giving some real world example and mentioning that the variable must be updated.

jroachgolf84 commented on providers/postgres/src/airflow/providers/postgres/triggers/postgres_cdc.py:
```diff
@@ -0,0 +1,127 @@
+
+                        max_iso = max_value.isoformat()
+
+                        last_value = Variable.get(self.state_key, default=None)
```
This doesn't need to change, but one thing that I did was add a `get_state` and `set_state` method that can eventually be subbed in. Right now, your `get_state` would just be a call to `Variable.get(...)`.

jroachgolf84 reviewed (changes requested):
I've been doing a bit more thinking about this, and I'm removing my approval (for now, again, this holds no actual weight as I'm not a committer). I'm concerned about the workflow with `Variable.get(...)` and having a Task downstream that `Variable.set()`'s the new value. I think folks are going to forget to do this, and this has real potential to bork their system.

jose-lehmkuhl committed:
feat(postgres): Add PostgresCDCEventTrigger for polling-based CDC

- Adds PostgresCDCEventTrigger to Postgres provider.
- Supports polling-based CDC using timestamp or version column.
- Integrates with AssetWatcher for triggering Airflow assets.
- Includes unit tests and example DAG.
- Adds state_key parameter for configurable state persistence.
- Adds version guard to new classes
- Adds trigger documentation
- Add PostgresCDCEventTrigger with recursive execution protection
